cpp/include/raft/matrix/gather.cuh

```diff
@@ -44,13 +47,13 @@ namespace raft::matrix {
  * @param  out          Pointer to the output matrix (assumed to be row-major)
  * @param  stream       CUDA stream to launch kernels within
  */
-template <typename MatrixIteratorT, typename MapIteratorT>
-void gather(const MatrixIteratorT in,
-            int D,
-            int N,
-            MapIteratorT map,
-            int map_length,
-            MatrixIteratorT out,
+template <typename InputIteratorT, typename MapIteratorT, typename OutputIteratorT, typename IndexT>
+void gather(const InputIteratorT in,
+            IndexT D,
+            IndexT N,
+            const MapIteratorT map,
+            IndexT map_length,
+            OutputIteratorT out,
             cudaStream_t stream)
```

Comment:
Not necessarily for this PR, but do you think it's feasible to implement a version of `gather` (and, maybe, `gather_if`) that takes the host pointer as the input `in`? e.g. load the `in` data into gpu in batches using `raft::copy` and then apply the `MapIteratorT` on it and write into `out` in a kernel.
I believe, it can be very handy for processing large datasets in algorithms like `ivf-pq`: copy a small part of a huge dataset with indices coming from random sampling (select the training set) or clustering (select the subset by the label).

Reply:
@achirkin Quick thoughts about this (definitely not for this PR):

- Approach 1: do multiple passes, only copy a row if the source is in the current batch. Problem: all the input dataset will be copied to the device at some point (though not concurrently), and also performance gets worse the more passes you have / the smaller the batches.
- Approach 2: gather on the host. Not completely stupid if the number of elements to gather is small.
- Approach 3: revert the map, sort it by source id, and do multiple passes, only looking at the part of the map that corresponds to the current input batch.